design/build-cache.md

```diff
@@ -0,0 +1,201 @@
+3. reject: Fails the build if there are any file conflicts.
+
+
+### Scoping
+
+Cache scoping levels can be expressed as `Job level < Event level < Pipeline level`. Job level scoping limits the cache to particular jobs, within a single event, within a single pipeline. Event level scoping limits the cache to a certain event within a single pipeline. Pipeline level scoping represents a globally scoped cache for the single pipeline, with the cache being available for all instances of the pipeline.
+
+#### [Tentative] Regarding commits, branches and scoping
+
+Caching needs to take into account commits and branches when performing scoping.
+
+Scoping a cache to commit level and/or not considering branches will lead to issues with caching in the following instances:
+
+- Build and create cache when deploying/build for commit in dev branch. Same cache gets matched when building commit for production branch, however production actually dynamically generates different resources/artifacts.
+- Have a single commit that gets deployed from a single PR or branch with multiple configurations.
+
+Scoping should therefore take into account:
+
+1. Branch
+2. Commit
+3. Hash signature of content to be cached
+
+#### Job level
+
+This level of caching explicitly links a cache to certain jobs.
+
+Use when:
+
+1. Injecting the cached resources into every job is not desirable.
+
+#### Event level
+
+Event level scoping is scoped to a single pipeline execution event. Cached data is visible and accessible to all jobs within a single execution of the pipeline.
+
+Executions of a pipeline for the same commit are viewed as different executions.
+
+Use when:
+
+1. Generated resources are time dependant
+2. Generated resources are only valid for single execution
+3. Generated resources are meant for deployment (Please note that triggering deployment with a detached job will create a new event and, by extension, cache.)
+
+#### Pipeline level
+
+Pipeline level scoping is scoped to a pipeline as a whole. Cached data is visible and accessible to all instances of the pipeline.
+
+Note: Concurrency for simultaneous builds need proper attention.
+
+Use when:
+
+1. Build dependencies take a long time to resolve
+2. Generated resources are meant for deployment (Risk prone, not recommended unless deleting deployed artifact from cache after deployment)
+
+### (Proposal) Local and remote caching
+
+Pipeline and event level caches are stored on the build cache service node. They are copied to the build node on pipeline setup, and back to the build cache on pipeline completion.
+
+Once copied to the build node, they are persisted as a shared volume, mounted onto the appropriate path in the build container. 
+
+This proposal is currently blocked by the nature of the backend implementation for nodes. The k8s executor for example, selects a completely random node for every job in the pipeline. Caching locally is therefore not viable at the moment.
```

Comment:
One possibility is to use NFS. With k8s one can setup nfs backed Persistent Volume Claim.

Reply:
- I'm not against implementing local storage for only a single backend (this actually won't be very different from the multiple node + NFS case, as NFS will be mounted to all the nodes as local directory). But the conflict you talked about could exist even in the single backend scenario, if multiple jobs in the same event run in parallel on the same node?
- Yes NFS will also be slow. I guess the main value would be the abstraction as if they are local disk volumes, and the reduced overhead. But will need testing to compare

Reply:
- 👍 I agree that r/w mount of NFS could be slow. Let's maybe focus on 1) the remote caching implementation and 2) local caching with a single backend implementation, until we have enough data to see how NFS performs.
- One other possibility is to copy the content in NFS to local files when build starts, and write them back when build finishes. So it would be identical to the remote caching proposal, but NFS will be used in place instead of HTTP
- If we go with implementing both the remote caching and a local disk caching, do we provide them as options that an SD admin can pick?
- I would still like to address the potential concurrency issue in the local storage caching proposal, where multiple builds can the same directory on the single execution node.

Reply:
- [ ] Create issue for testing performance of NFS backed shared volume vs copying for every job.